File: app/src/main/java/com/example/bookapp/viewModels/CommentsRepository.java
package com.example.bookapp.viewModels;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import com.example.dataLayer.interfaces.CommentsInterface;
import com.example.bookapp.models.Comment;
import com.example.dataLayer.dataObjectsToSerialize.SerializeComment;

import java.util.ArrayList;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

class CommentsRepository {
    private CommentsInterface commentsInterface;
    private static CommentsRepository instance;
    private MutableLiveData<ArrayList<Comment>> postComments;

    public static synchronized CommentsRepository getInstance(@NonNull Retrofit retrofit) {
        if (instance == null) {
            instance = new CommentsRepository(retrofit);
        }
        return instance;
    }

    private CommentsRepository(@NonNull Retrofit retrofit) {
        commentsInterface = retrofit.create(CommentsInterface.class);
    }

    void uploadComment(@NonNull SerializeComment comment) {
        commentsInterface.uploadComment(true, comment).enqueue(new Callback<Comment>() {
            @Override
            public void onResponse(@NonNull Call<Comment> call, @NonNull Response<Comment> response) {
                //the postComments and the value of post comments
                //should not be null but we should check
                addCommentAndNotify(response.body());
            }

            @Override
            public void onFailure(@NonNull Call<Comment> call, @NonNull Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void addCommentAndNotify(Comment comment) {
        if (postComments != null && postComments.getValue() != null) {
            ArrayList<Comment> newData = new ArrayList<>(postComments.getValue());
            newData.add(comment);
            postComments.setValue(newData);
        }
    }

    @NonNull
    MutableLiveData<ArrayList<Comment>> fetchCommentsForPost(int postID) {
        postComments = new MutableLiveData<>();

        commentsInterface.fetchCommentsByPostID(postID, true).enqueue(new Callback<ArrayList<Comment>>() {
            @Override
            public void onResponse(@NonNull Call<ArrayList<Comment>> call, @NonNull Response<ArrayList<Comment>> response) {
                postComments.setValue(response.body());
            }

            @Override
            public void onFailure(@NonNull Call<ArrayList<Comment>> call, @NonNull Throwable t) {
                t.printStackTrace();
            }
        });
        return postComments;
    }
}

File: app/src/main/java/com/example/bookapp/viewModels/CommentsRepository.kt
package com.example.bookapp.viewModels

import android.net.ConnectivityManager
import androidx.lifecycle.LiveData
import androidx.lifecycle.liveData
import com.example.bookapp.models.Post
import com.example.dataLayer.dataMappers.CommentMapper
import com.example.dataLayer.interfaces.CommentRepoInterface
import com.example.dataLayer.interfaces.dao.RoomCommentDao
import com.example.dataLayer.models.PostWithComments
import com.example.dataLayer.models.serialization.SerializeComment
import com.example.dataLayer.repositories.OperationStatus
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.InternalCoroutinesApi
import kotlinx.coroutines.launch
import javax.inject.Inject

@InternalCoroutinesApi
class CommentsRepository @Inject constructor(private val connectivityManager: ConnectivityManager,
                                             private val commentDao: RoomCommentDao,
                                             private val repo: CommentRepoInterface) {


    fun getCommentsForPost(post: Post): LiveData<PostWithComments> = liveData {
     //   emitSource(commentDao.getAllPostComments(post.id))
        if (connectivityManager.activeNetwork != null) {
            fetchCommentsForPost(post)
        }
    }


    fun uploadComment(comment: SerializeComment) = liveData {
        emit(OperationStatus.ONGOING)
        try {
            val serverResponse = repo.uploadComment(comment)
            val commentID = serverResponse.message.toIntOrNull()
            commentID?.let {
                val fetchedComment = repo.fetchCommentById(it)
                commentDao.insertComment(CommentMapper.mapToDomainObject(fetchedComment))
                emit(OperationStatus.FINISHED)
            }

        } catch (e: Exception) {
            e.printStackTrace()
            emit(OperationStatus.FAILED)
        }
    }


    private suspend fun fetchCommentsForPost(post: Post) {
        try {
           // val fetchedComments = repo.fetchCommentsForPost(post.id)
           // commentDao.insertComments(fetchedComments.map { CommentMapper.mapToDomainObject(it) })
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

}